indexer/name_index.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2016 The Vulcan Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package indexer
+
+import "sync"
+
+// NameIndex groups metrics by the label "__name__"
+type NameIndex struct {
+	entries map[string]*Index
+	m       sync.RWMutex
+}
+
+// NewNameIndex creates a NameIndex.
+func NewNameIndex() *NameIndex {
+	return &NameIndex{
+		entries: map[string]*Index{},
+	}
+}
+
+// Add inserts an id and its labels into the index. It is expected that the labels for
+// a given ID will always be the same.
+func (ni *NameIndex) Add(id string, labels map[string]string) {
+	name := labels["__name__"]
```

Comment:
how do you feel about importing [this](https://github.com/prometheus/common/blob/master/model/labels.go#L34) instead to add some resiliency to upstream changes?